Coxeter integration

VincentRouvreau opened the pull request.

mglisse commented on src/Coxeter_triangulation/include/gudhi/IO/build_mesh_from_cell_complex.h:
```diff
@@ -0,0 +1,167 @@
+}
+
+template <class Cell_complex>
+Mesh_medit build_mesh_from_cell_complex(const Cell_complex& cell_complex,
```
No documentation?

VincentRouvreau replied:
doc review: fix `build_mesh_from_cell_complex` and `output_meshes_to_medit` doxygen issue on af94bf9d

mglisse commented on src/Coxeter_triangulation/include/gudhi/IO/output_meshes_to_medit.h:
```diff
@@ -0,0 +1,151 @@
+ *  @param[in] meshes A pack of meshes to be specified separated by commas.
+ */
+template <typename... Meshes>
+void output_meshes_to_medit(std::size_t amb_d, std::string file_name, const Meshes&... meshes) {
```
This one has a doc comment, but I don't see it in the generated doc.

VincentRouvreau replied:
doc review: fix `build_mesh_from_cell_complex` and `output_meshes_to_medit` doxygen issue on af94bf9d

VincentRouvreau commented:
@mglisse I also specified that the cell complex was only for Coxeter on 29fa56b

mglisse reviewed (commented):
Hasse_diagram_cell is not advertised from the front page, that's good, but I don't see a warning in its doc that it is experimental, just for coxeter, or something like that to discourage someone who would casually notice it from trying to use it for something else.

mglisse commented:
I see some suspicious diff in ext/ submodules.

mglisse commented on src/common/include/gudhi/math.h:
```diff
@@ -0,0 +1,23 @@
+namespace Gudhi {
+
+// In wait of C++20 std::numbers::pi with #include <numbers>
+static constexpr double PI = boost::math::constants::pi<double>();
```
I think in another branch we said we might as well use the boost constants directly, without a gudhi wrapper.

VincentRouvreau replied:
Done on https://github.com/GUDHI/gudhi-devel/pull/397/commits/3a06eff2f6345133270ddc50b625139761c06df6

mglisse commented:
The list of headers is

gudhi/Cell_complex.h
gudhi/Cell_complex/Hasse_diagram_cell.h
gudhi/Coxeter_triangulation.h
gudhi/Freudenthal_triangulation.h
gudhi/Functions/Cartesian_product.h
gudhi/Functions/Constant_function.h
gudhi/Functions/Embed_in_Rd.h
gudhi/Functions/Function_Sm_in_Rd.h
gudhi/Functions/Function_affine_plane_in_Rd.h
gudhi/Functions/Function_chair_in_R3.h
gudhi/Functions/Function_iron_in_R3.h
gudhi/Functions/Function_lemniscate_revolution_in_R3.h
gudhi/Functions/Function_moment_curve_in_Rd.h
gudhi/Functions/Function_torus_in_R3.h
gudhi/Functions/Function_whitney_umbrella_in_R3.h
gudhi/Functions/Linear_transformation.h
gudhi/Functions/Negation.h
gudhi/Functions/PL_approximation.h
gudhi/Functions/Translate.h
gudhi/Functions/random_orthogonal_matrix.h
gudhi/IO/Mesh_medit.h
gudhi/IO/build_mesh_from_cell_complex.h
gudhi/IO/output_debug_traces_to_html.h
gudhi/IO/output_meshes_to_medit.h
gudhi/Implicit_manifold_intersection_oracle.h
gudhi/Manifold_tracing.h
gudhi/Permutahedral_representation.h
gudhi/Permutahedral_representation/Combination_iterator.h
gudhi/Permutahedral_representation/Integer_combination_iterator.h
gudhi/Permutahedral_representation/Ordered_set_partition_iterator.h
gudhi/Permutahedral_representation/Permutahedral_representation_iterators.h
gudhi/Permutahedral_representation/Permutation_iterator.h
gudhi/Permutahedral_representation/Set_partition_iterator.h
gudhi/Permutahedral_representation/Simplex_comparator.h
gudhi/Permutahedral_representation/Size_range.h
gudhi/Permutahedral_representation/face_from_indices.h
gudhi/Query_result.h

Some are in a subdirectory Permutahedral_representation, that's good. The subdirectory Functions has a name that seems a bit too generic, but maybe it isn't so bad. Query_result.h seems very generic for something at the top level, and I don't think users ever have to include it directly. Maybe move it to some subdirectory? We may also want to avoid Cell_complex at the top level while it is experimental and restricted to coxeter.

VincentRouvreau commented:
> Hasse_diagram_cell is not advertised from the front page, that's good, but I don't see a warning in its doc that it is experimental, just for coxeter, or something like that to discourage someone who would casually notice it from trying to use it for something else.

You are right, I forget this one. Done on 44b7b7ce

VincentRouvreau commented on src/Coxeter_triangulation/include/gudhi/Cell_complex.h:
```diff
@@ -0,0 +1,344 @@
+    if (map_it == simplex_cell_map.end()) {
+      hasse_cells_.push_back(new Hasse_cell(is_boundary, cell_d));
+      Hasse_cell* new_cell = hasse_cells_.back();
+      simplex_cell_map.emplace(std::make_pair(simplex, new_cell));
```
I really need to acquire this reflex. Done on 0828b90f

VincentRouvreau commented:
> The list of headers is
> 
> gudhi/Cell_complex.h
> gudhi/Cell_complex/Hasse_diagram_cell.h
> gudhi/Coxeter_triangulation.h
> gudhi/Freudenthal_triangulation.h
> gudhi/Functions/Cartesian_product.h
> gudhi/Functions/Constant_function.h
> gudhi/Functions/Embed_in_Rd.h
> gudhi/Functions/Function_Sm_in_Rd.h
> gudhi/Functions/Function_affine_plane_in_Rd.h
> gudhi/Functions/Function_chair_in_R3.h
> gudhi/Functions/Function_iron_in_R3.h
> gudhi/Functions/Function_lemniscate_revolution_in_R3.h
> gudhi/Functions/Function_moment_curve_in_Rd.h
> gudhi/Functions/Function_torus_in_R3.h
> gudhi/Functions/Function_whitney_umbrella_in_R3.h
> gudhi/Functions/Linear_transformation.h
> gudhi/Functions/Negation.h
> gudhi/Functions/PL_approximation.h
> gudhi/Functions/Translate.h
> gudhi/Functions/random_orthogonal_matrix.h
> gudhi/IO/Mesh_medit.h
> gudhi/IO/build_mesh_from_cell_complex.h
> gudhi/IO/output_debug_traces_to_html.h
> gudhi/IO/output_meshes_to_medit.h
> gudhi/Implicit_manifold_intersection_oracle.h
> gudhi/Manifold_tracing.h
> gudhi/Permutahedral_representation.h
> gudhi/Permutahedral_representation/Combination_iterator.h
> gudhi/Permutahedral_representation/Integer_combination_iterator.h
> gudhi/Permutahedral_representation/Ordered_set_partition_iterator.h
> gudhi/Permutahedral_representation/Permutahedral_representation_iterators.h
> gudhi/Permutahedral_representation/Permutation_iterator.h
> gudhi/Permutahedral_representation/Set_partition_iterator.h
> gudhi/Permutahedral_representation/Simplex_comparator.h
> gudhi/Permutahedral_representation/Size_range.h
> gudhi/Permutahedral_representation/face_from_indices.h
> gudhi/Query_result.h
> 
> Some are in a subdirectory Permutahedral_representation, that's good. The subdirectory Functions has a name that seems a bit too generic, but maybe it isn't so bad. Query_result.h seems very generic for something at the top level, and I don't think users ever have to include it directly. Maybe move it to some subdirectory? We may also want to avoid Cell_complex at the top level while it is experimental and restricted to coxeter.

I moved Query_results.h in a `Coxeter_triangulation` directory and I also moved all `Cell_complex` implementation in it on bc9411b2

VincentRouvreau commented:
> I see some suspicious diff in ext/ submodules.

I will fix it by merging only what is required in another when this PR will be approved. If it is ok for you. Don't know what happened here...
